### PR TITLE
[FrameworkBundle] Deprecate renderView() in favor of renderTemplate()

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | -
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 1]
```diff
@@ -241,20 +241,30 @@ protected function denyAccessUnlessGranted($attributes, $subject = null, string
...
+    {
+        @trigger_error('Since symfony/framework-bundle 5.1: "renderView()" method is deprecated. Use "renderTemplate()" instead.', E_USER_DEPRECATED);
+
+        return $this->renderTemplate($view, $parameters);
+    }
+
+    /**
+     * Returns a rendered template.
+     */
+    protected function renderTemplate(string $templatePath, array $parameters = []): string
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: should be named `$templateName`, not `$templatePath`. That's not a path.

**Comment by stof** [comment]
@nicolas-grekas I reverted the milestone. We cannot introduce a new deprecation in 5.0.x

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 2]
```diff
@@ -241,20 +241,30 @@ protected function denyAccessUnlessGranted($attributes, $subject = null, string
      * Returns a rendered view.
      */
     protected function renderView(string $view, array $parameters = []): string
+    {
+        @trigger_error('Since symfony/framework-bundle 5.1: "renderView()" method is deprecated. Use "renderTemplate()" instead.', E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: trigger_deprecation() + missing annotation

**Review by nicolas-grekas — COMMENTED** [review]
Do we consider this is worth the migration cost?

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: oh also: changelog+upgrade files should be updated

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php (hunk 1 above)
Comment (javiereguiluz) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: My reasoning was:

```
blog/index.html.twig

* Template name: 'index.html.twig'
* Template path: 'blog/index.html.twig'
* Template file path: '<project>/templates/blog/index.html.twig'
```

**Comment by javiereguiluz** [comment]
About the migration cost ... I would never change the `render()` method because *everybody* uses it ... but `renderView()` is much less popular (in my experience and from what I've seen out there: render() is used more than 100 times more often that renderView() ... but this is only my experience).

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: that's the wrong expectation. What Twig calls the template name is that `blog/index.html.twig` (or `@App/index.hml.twig`). Twig never deals with `index.html.twig` here.

And there is nothing forcing a template name to look like a path (the FilesystemLoader is the one bringing such behavior, but namespaces already causes a difference here).

**Comment by dmaicher** [comment]
> About the migration cost ... I would never change the `render()` method because _everybody_ uses it ... but `renderView()` is much less popular (in my experience and from what I've seen out there: render() is used more than 100 times more often that renderView() ... but this is only my experience).

Seems about right. Not exactly a factor of 100 but here some numbers from my biggest Symfony 4.4 work project:

```
$ grep -r --include *Controller.php  'this->render(' src/ vendor/ | wc -l
180

$ grep -r --include *Controller.php  'this->renderView(' src/ vendor/ | wc -l
14

**Comment by michaljusiega** [comment]
Another point. MVC division: Model, view, controller. Most programmers associate with such an approach that the controller generates a view, not a template.

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 3]
```diff
@@ -242,7 +242,7 @@ protected function denyAccessUnlessGranted($attributes, $subject = null, string
      */
     protected function renderView(string $view, array $parameters = []): string
     {
-        @trigger_error('Since symfony/framework-bundle 5.1: "renderView()" method is deprecated. Use "renderTemplate()" instead.', E_USER_DEPRECATED);
+        @trigger_deprecation('Since symfony/framework-bundle 5.1: "renderView()" method is deprecated. Use "renderTemplate()" instead.', E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: `s/@//`

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: well actually that's broken, you'll figure out :)

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: (don't miss the annotation also)

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 4]
```diff
@@ -241,20 +241,30 @@ protected function denyAccessUnlessGranted($attributes, $subject = null, string
      * Returns a rendered view.
      */
     protected function renderView(string $view, array $parameters = []): string
+    {
+        trigger_deprecation('Since symfony/framework-bundle 5.1: "renderView()" method is deprecated. Use "renderTemplate()" instead.', E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: The "renderView()" method [...]?

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]: that's not the signature of the function, and the annotation is missing

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
(just to confirm this needs a few more hops)

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 5]
```diff
@@ -239,22 +239,34 @@ protected function denyAccessUnlessGranted($attributes, $subject = null, string
 
     /**
      * Returns a rendered view.
+     *
+     * @deprecated since Symfony 5.1, use renderTemplate() instead.
      */
     protected function renderView(string $view, array $parameters = []): string
+    {
+        trigger_deprecation('symfony/framework-bundle', '5.1', 'The "renderView()" method is deprecated, use "renderTemplate()" instead.', E_USER_DEPRECATED);
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:247]: 3rd arg: `'The "%s" method is [...]`
4th arg: `__METHOD__`

Comment (javiereguiluz) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:247]: Thanks! Changes done.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Comment by fabpot** [comment]
Actually, that's a hard BC break. What if one already has a `renderTemplate()` method in their controllers? ... well, it happens that EasyAdmin does this, and it is broken now. Shall we revert?

**Comment by nicolas-grekas** [comment]
Sure, let's revert if it needs more thoughts.

**Comment by javiereguiluz** [comment]
@fabpot I hadn't thought about that ... but then, we'll never be able to add any new method to AbstractController ... because someone may have already defined it 🤔

**Comment by fabpot** [comment]
@javiereguiluz That's correct :(

**Comment by fabpot** [comment]
reverted

**Comment by javiereguiluz** [comment]
A possible solution would be to deprecate defining custom `renderTemplate()` methods and adding it in Symfony 6.0:

```php
public function renderView(...)
{
    if (method_exists($this, 'renderTemplate')) {
        @trigger_deprecation('
            You define a custom renderTemplate() method in your controllers,
            but Symfony will add that method in Symfony 6.0, so you should
            rename your method.
        ');
    }

    // ...
}
```

**Pull request merged**